src/track/beatutils.cpp

```diff
@@ -12,37 +12,314 @@
+    // last change sentinal does happens in the right place.
+    stableTemposByPosition.remove(stableTemposByPosition.last());
+    stableTemposByPosition[anchoredBeats.size() -1] = stableTemposByPosition[0];
+    rawBeats = anchoredBeats;
+}
+
+QVector<double> BeatUtils::FixBeatmap(
+        QVector<double>& rawBeats, int sampleRate, double minBpm, double maxBpm) {
+    
+    QMap<double, int> tempoFrequency;
+    QList<double> tempoList = computeWindowedBpmsAndFrequencyHistogram(
+            rawBeats, 2, 1, sampleRate, &tempoFrequency);
+    auto stableTemposByPosition = findTempoChanges(tempoFrequency, tempoList);
+    RemoveSmallArrhythmic(rawBeats, sampleRate, stableTemposByPosition);
+    QVector<double> fixedBeats;
+    fixedBeats.reserve(rawBeats.size());
+    auto tempoChanges = stableTemposByPosition.keys();
+    for (int lastTempoChage = 1; 
+                lastTempoChage < tempoChanges.size();
+                lastTempoChage++) {
+
+        int beatStart = tempoChanges[lastTempoChage -1];
+        // when adding the beats in a new tempo in the previous iteration
+        //  we might have already added beats at this positions so skip then...
+        if (fixedBeats.size() > 0) {
+            while (rawBeats[beatStart] <= fixedBeats.last()) {
+                beatStart += 1;
+                if (beatStart == rawBeats.size()) {
+                    break;
+                }
+            }
+        }
+        int beatEnd = tempoChanges[lastTempoChage];
+        // in this case we have already considered all these beats..
+        if (beatStart >= beatEnd) {
+            continue;
+        }
+        int partLenght = beatEnd - beatStart;
+        double leftRighDiff = 1.0;
+        // here we detect if the segment has a constant tempo or not
+        if (partLenght >= kBeatsToCountTempo * 2) {
+            int middle = partLenght / 2;            
+            auto beatsAtLeft = QVector<double>::fromStdVector(std::vector<double>(
+                    rawBeats.begin() + beatStart, rawBeats.begin() + beatStart + middle));
+            auto beatsAtRight = QVector<double>::fromStdVector(std::vector<double>(
+                    rawBeats.begin() + beatStart + middle, rawBeats.begin() + beatEnd));
+            QMap<double, int> leftTempoFrequency;
+            auto temposLeft = computeWindowedBpmsAndFrequencyHistogram(
+                    beatsAtLeft, kBeatsToCountTempo, 1, sampleRate, &leftTempoFrequency);
+            QMap<double, int> rightTempoFrequency;
+            auto temposRight = computeWindowedBpmsAndFrequencyHistogram(
+                        beatsAtRight, kBeatsToCountTempo, 1, sampleRate, &rightTempoFrequency);
+            double modeAtLeft = BeatStatistics::mode(leftTempoFrequency);
+            double modeAtRight = BeatStatistics::mode(rightTempoFrequency);
+            leftRighDiff = fabs(modeAtLeft - modeAtRight);
+        }
+        // const tempo make const grid
+        if (leftRighDiff < kMaxDiffSameBpm) {
```

Comment:
I did not get the consideration for this test. 
Please describe in a code comment.

Reply:
done